pkg/extprom/http/metrics.go

```diff
@@ -10,31 +10,40 @@ import (
 
 type defaultMetrics struct {
 	requestDuration      *prometheus.HistogramVec
-	requestSize          *prometheus.SummaryVec
+	requestSize          *prometheus.HistogramVec
 	requestsTotal        *prometheus.CounterVec
-	responseSize         *prometheus.SummaryVec
+	responseSize         *prometheus.HistogramVec
 	inflightHTTPRequests *prometheus.GaugeVec
 }
 
-func newDefaultMetrics(reg prometheus.Registerer, buckets []float64, extraLabels []string) *defaultMetrics {
-	if buckets == nil {
-		buckets = []float64{0.001, 0.01, 0.1, 0.3, 0.6, 1, 3, 6, 9, 20, 30, 60, 90, 120, 240, 360, 720}
+func newDefaultMetrics(reg prometheus.Registerer, durationBuckets []float64, extraLabels []string) *defaultMetrics {
+	if durationBuckets == nil {
+		durationBuckets = []float64{0.001, 0.01, 0.1, 0.3, 0.6, 1, 3, 6, 9, 20, 30, 60, 90, 120, 240, 360, 720}
 	}
 
+	bytesBuckets := prometheus.ExponentialBuckets(64, 2, 10)
+	bucketFactor := 1.1
+	maxBuckets := uint32(100)
+
 	return &defaultMetrics{
 		requestDuration: promauto.With(reg).NewHistogramVec(
 			prometheus.HistogramOpts{
-				Name:    "http_request_duration_seconds",
-				Help:    "Tracks the latencies for HTTP requests.",
-				Buckets: buckets,
+				Name:                           "http_request_duration_seconds",
+				Help:                           "Tracks the latencies for HTTP requests.",
+				Buckets:                        durationBuckets,
+				NativeHistogramBucketFactor:    bucketFactor,
+				NativeHistogramMaxBucketNumber: maxBuckets,
 			},
 			append([]string{"code", "handler", "method"}, extraLabels...),
 		),
 
-		requestSize: promauto.With(reg).NewSummaryVec(
-			prometheus.SummaryOpts{
-				Name: "http_request_size_bytes",
-				Help: "Tracks the size of HTTP requests.",
+		requestSize: promauto.With(reg).NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:                           "http_request_size_bytes",
+				Help:                           "Tracks the size of HTTP requests.",
+				Buckets:                        bytesBuckets,
+				NativeHistogramBucketFactor:    bucketFactor,
+				NativeHistogramMaxBucketNumber: maxBuckets,
```

Comment:
@rabenhorst does this mean these two metrics became native histograms? What are the consequences of this in terms of backwards compatibility with folks that scrape metrics from their Thanos version with a version of Prometheus that still doesn't support native histograms?

Reply:
Native histograms are quite a recent addition to Prometheus, still considered experimental, [and has plenty of issues reported in Prometheus](https://github.com/prometheus/prometheus/issues?q=is%3Aissue+is%3Aopen+native+histogram). We are forcing people that want to observe these metrics to use an experimental feature with plenty of bugs in favor of the classic histograms. Maybe we should move back to classic histograms and use the native ones when they are more stable and widespread?

Reply:
Ah, I learned that native histograms still export the classic histogram metrics. It's all fine then, I guess. 😄

Reply:
Yes. So native histograms are only scraped when Prometheus  (=> v2.4.0) and the client ( >= v1.14.0) support it, and it is enabled for both sides (flag for Prometheus and `NativeHistogramBucketFactor` > 1 for client). Otherwise Prometheus will scrape conventional histograms.